server/src/main/java/org/apache/cloudstack/vm/lease/VMLeaseManagerImpl.java

```diff
@@ -0,0 +1,302 @@
+                    logger.warn("Catch throwable in VM lease scheduler ", t);
+                }
+            }
+        };
+
+        vmLeaseTimer = new Timer("VMLeasePollTask");
+        vmLeaseTimer.scheduleAtFixedRate(schedulerPollTask, 5_000L, InstanceLeaseSchedulerInterval.value() * 1000L);
+
+        vmLeaseAlterTimer = new Timer("VMLeaseAlertPollTask");
+        vmLeaseAlterTimer.scheduleAtFixedRate(leaseAlterSchedulerTask, 5_000L, InstanceLeaseAlertSchedule.value() * 1000)
+        ;
+        return true;
+    }
+
+    protected void alert() {
+        // as feature is disabled, no action is required
+        if (!InstanceLeaseEnabled.value()) {
+            return;
+        }
+
+        GlobalLock scanLock = GlobalLock.getInternLock("VMLeaseAlertScheduler");
+        try {
+            if (scanLock.lock(ACQUIRE_GLOBAL_LOCK_TIMEOUT_FOR_COOPERATION)) {
+                try {
+                    List<UserVmJoinVO> leaseExpiringForInstances = userVmJoinDao.listLeaseInstancesExpiringInDays(InstanceLeaseAlertStartsAt.value().intValue());
+                    for (UserVmJoinVO instance : leaseExpiringForInstances) {
+                        alertManager.sendAlert(AlertManager.AlertType.ALERT_TYPE_USERVM, instance.getDataCenterId(), instance.getPodId(),
+                                "Lease expiring for instance id: " + instance.getUuid(), "Lease expiring for instance");
+                    }
+                } finally {
+                    scanLock.unlock();
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+    }
+
+    @Override
+    public void poll(Date timestamp) {
+        Date currentTimestamp = DateUtils.round(timestamp, Calendar.MINUTE);
+        String displayTime = DateUtil.displayDateInTimezone(DateUtil.GMT_TIMEZONE, currentTimestamp);
+        logger.debug("VMLeaseScheduler.poll is being called at {}", displayTime);
+
+        if (!InstanceLeaseEnabled.value()) {
+            logger.debug("Instance lease feature is disabled, no action is required");
+            return;
+        }
+
+        GlobalLock scanLock = GlobalLock.getInternLock("VMLeaseScheduler");
+        try {
+            if (scanLock.lock(ACQUIRE_GLOBAL_LOCK_TIMEOUT_FOR_COOPERATION)) {
+                try {
+                    reallyRun();
+                } finally {
+                    scanLock.unlock();
+                }
+            }
+        } finally {
+            scanLock.releaseRef();
+        }
+    }
+
+    protected void reallyRun() {
+        // fetch user_instances having leaseDuration configured and has expired
+        List<UserVmJoinVO> leaseExpiredInstances = userVmJoinDao.listLeaseExpiredInstances();
+        List<Long> actionableInstanceIds = new ArrayList<>();
+        // iterate over and skip ones with delete protection
+        for (UserVmJoinVO userVmVO : leaseExpiredInstances) {
+            if (userVmVO.isDeleteProtection() != null && userVmVO.isDeleteProtection()) {
```

Comment:
only when lease action is DESTROY, instance with lease expired should be ignored (this behavior needs to be documented)